### PR TITLE
Schema types for action and trigger submission

### PR DESCRIPTION
**Pull request by gusingh-r7**
Add input and output schema types for action and trigger submission

### CODE REVIEW

**Comment by gusingh-r7** [comment]
http://localhost:10001/api

```
{
    "paths": {
        "/api": {
            "get": {
                "summary": "Get API spec details",
                "description": "Get Swagger v2.0 API Specification",
                "parameters": [
                    {
                        "in": "query",
                        "name": "format",
                        "type": "string",
                        "description": "Format to return swagger spec; defaults to JSON",
                        "enum": [
                            "json",
                            "yaml"
                        ]
                    }
                ],
                "responses": {
                    "200": {
                        "description": "Swagger Specification to be returned",
                        "schema": {
                            "type": "object"
                        }
                    },
                    "422": {
                        "description": "The specified format is not supported"
                    }
                }
            }
        },
        "/info": {
            "get": {
                "summary": "Get plugin details",
                "description": "Get InsightConnect plugin details",
                "responses": {
                    "200": {
                        "description": "InsightConnect Plugin Information to be returned",
                        "schema": {
                            "$ref": "#/definitions/PluginInfo"
                        }
                    }
                }
            }
        },
        "/spec": {
            "get": {
                "summary": "Get plugin spec details",
                "description": "Get plugin specification",
                "parameters": [
                    {
                        "in": "query",
                        "name": "format",
                        "type": "string",
                        "description": "Format to return plugin spec; defaults to JSON",
                        "enum": [
                            "json",
                            "yaml"
                        ]
                    }
                ],
                "responses": {
                    "200": {
                        "description": "Plugin specification to be returned",
                        "schema": {
                            "type": "object"
                        }
                    },
                    "422": {
                        "description": "The specified format is not supported"
                    }
                }
            }
        },
        "/actions": {
            "get": {
                "summary": "Get list of plugin actions",
                "description": "Get InsightConnect plugin all actions",
                "responses": {
                    "200": {
                        "description": "InsightConnect Plugin actions list to be returned",
                        "schema": {
                            "type": "array"
                        }
                    }
                }
            }
        },
        "/triggers": {
            "get": {
                "summary": "Get list of plugin triggers",
                "description": "Get InsightConnect plugin all triggers",
                "responses": {
                    "200": {
                        "description": "InsightConnect Plugin triggers list to be returned",
                        "schema": {
                            "type": "array"
                        }
                    }
                }
            }
        },
        "/status": {
            "get": {
                "summary": "Get web service status",
                "description": "Get web service status",
                "responses": {
                    "200": {
                        "description": "Status to be returned",
                        "schema": {
                            "type": "object"
                        }
                    }
                }
            }
        },
        "/actions/{name}": {
            "post": {
                "summary": "Run an action",
                "description": "Run an action",
                "parameters": [
                    {
                        "name": "name",
                        "in": "path",
                        "description": "Name of the action",
                        "type": "string",
                        "required": true
                    },
                    {
                        "in": "formData",
                        "name": "version",
                        "required": true,
                        "type": "string"
                    },
                    {
                        "in": "formData",
                        "name": "type",
                        "required": true,
                        "type": "string",
                        "enum": [
                            "action_event",
                            "trigger_event"
                        ]
                    },
                    {
                        "in": "formData",
                        "name": "body",
                        "required": true,
                        "$ref": "#/definitions/ActionTriggerInputBody"
                    }
                ],
                "responses": {
                    "200": {
                        "description": "Action output to be returned",
                        "schema": {
                            "$ref": "#/definitions/ActionTriggerOutput"
                        }
                    },
                    "400": {
                        "description": "Bad request"
                    },
                    "500": {
                        "description": "Unexpected error"
                    }
                }
            }
        },
        "/actions/{name}/test": {
            "post": {
                "summary": "Run action test",
                "description": "Run action test",
                "parameters": [
                    {
                        "name": "name",
                        "in": "path",
                        "description": "Name of the action",
                        "type": "string",
                        "required": true
                    },
                    {
                        "in": "formData",
                        "name": "version",
                        "required": true,
                        "type": "string"
                    },
                    {
                        "in": "formData",
                        "name": "type",
                        "required": true,
                        "type": "string",
                        "enum": [
                            "action_event",
                            "trigger_event"
                        ]
                    },
                    {
                        "in": "formData",
                        "name": "body",
                        "required": true,
                        "$ref": "#/definitions/ActionTriggerInputBody"
                    }
                ],
                "responses": {
                    "200": {
                        "description": "Action test output to be returned",
                        "schema": {
                            "$ref": "#/definitions/ActionTriggerOutput"
                        }
                    },
                    "400": {
                        "description": "Bad request"
                    },
                    "500": {
                        "description": "Unexpected error"
                    }
                }
            }
        },
        "/triggers/{name}/test": {
            "post": {
                "summary": "Run trigger test",
                "description": "Run trigger test",
                "parameters": [
                    {
                        "name": "name",
                        "in": "path",
                        "description": "Name of the trigger",
                        "type": "string",
                        "required": true
                    },
                    {
                        "in": "formData",
                        "name": "version",
                        "required": true,
                        "type": "string"
                    },
                    {
                        "in": "formData",
                        "name": "type",
                        "required": true,
                        "type": "string",
                        "enum": [
                            "action_event",
                            "trigger_event"
                        ]
                    },
                    {
                        "in": "formData",
                        "name": "body",
                        "required": true,
                        "$ref": "#/definitions/ActionTriggerInputBody"
                    }
                ],
                "responses": {
                    "200": {
                        "description": "Trigger test output to be returned",
                        "schema": {
                            "$ref": "#/definitions/ActionTriggerOutput"
                        }
                    },
                    "400": {
                        "description": "Bad request"
                    },
                    "500": {
                        "description": "Unexpected error"
                    }
                }
            }
        }
    },
    "info": {
        "title": "Runtime API",
        "version": "1.0"
    },
    "swagger": "2.0",
    "definitions": {
        "PluginInfo": {
            "type": "object",
            "properties": {
                "number_of_workers": {
                    "type": "integer",
                    "format": "int32"
                },
                "version": {
                    "type": "string"
                },
                "name": {
                    "type": "string"
                },
                "description": {
                    "type": "string"
                },
                "threads": {
                    "type": "integer",
                    "format": "int32"
                },
                "vendor": {
                    "type": "string"
                }
            }
        },
        "ActionTriggerOutputBody": {
            "type": "object",
            "properties": {
                "meta": {
                    "type": "object"
                },
                "log": {
                    "type": "string"
                },
                "output": {
                    "type": "object"
                },
                "status": {
                    "type": "string"
                }
            },
            "required": [
                "log",
                "meta",
                "output",
                "status"
            ]
        },
        "ActionTriggerOutput": {
            "type": "object",
            "properties": {
                "version": {
                    "type": "string"
                },
                "type": {
                    "type": "string",
                    "enum": [
                        "action_event",
                        "trigger_event"
                    ]
                },
                "body": {
                    "$ref": "#/definitions/ActionTriggerOutputBody"
                }
            },
            "required": [
                "body",
                "type",
                "version"
            ]
        },
        "ActionTriggerInputBody": {
            "type": "object",
            "properties": {
                "input": {
                    "type": "object"
                },
                "connection": {
                    "type": "object"
                },
                "action": {
                    "type": "string"
                }
            },
            "required": [
                "action",
                "connection",
                "input"
            ]
        },
        "ActionTriggerInput": {
            "type": "object",
            "properties": {
                "version": {
                    "type": "string"
                },
                "type": {
                    "type": "string",
                    "enum": [
                        "action_event",
                        "trigger_event"
                    ]
                },
                "body": {
                    "$ref": "#/definitions/ActionTriggerInputBody"
                }
            },
            "required": [
                "body",
                "type",
                "version"
            ]
        }
    }
}
```

http://localhost:10001/actions/run/test
```
{
    "body": {
        "log": "rapid7/Python 3 Script:2.0.1. Step name: run\n[*] Performing Python version check...\n\nPython 3.7.4\n\n",
        "meta": {},
        "output": {},
        "status": "ok"
    },
    "type": "action_event",
    "version": "v1"
}
```

http://localhost:10001/actions/run
```
{
    "body": {
        "log": "rapid7/Python 3 Script:2.0.1. Step name: run\nInput: (below)\n\n{'name': 'gur'}\n\nFunction: (below)\n\ndef run(params={}):\n return {'name': params['name'] }\n\n",
        "meta": {},
        "output": {
            "name": "gur"
        },
        "status": "ok"
    },
    "type": "action_event",
    "version": "v1"
}
```

File: komand/server.py [hunk 1]
```diff
@@ -29,6 +30,31 @@ class PluginInfoSchema(Schema):
...
+class ActionTriggerOutputBodySchema(Schema):
+    log = fields.Str(required=True)
+    meta = fields.Dict(required=True)
+    output = fields.Dict(required=True)
+    status = fields.Str(required=True)
+
+
+class ActionTriggerOutputSchema(Schema):
+    body = fields.Nested(ActionTriggerOutputBodySchema, required=True)
+    type = fields.Str(required=True, validate=OneOf(['action_event', 'trigger_event']))
```
Comment (zyoutz-r7) [inline komand/server.py]: Nice 👍

File: komand/server.py [hunk 2]
```diff
@@ -164,8 +202,14 @@ def action_test(name):
             input_message = request.get_json(force=True)
             self.logger.debug('Request input: %s', input_message)
             PluginServer.validate_action_trigger(input_message, name, 'action')
-            output = self.run_action_trigger(input_message, True)
-            return output
+            response = self.run_action_trigger(input_message, True)
+            action_response = json.loads(response.response[0])
```
Comment (zyoutz-r7) [inline komand/server.py]: What requires returning index 0?  I thought we returned a map/object here?

**Comment by zyoutz-r7** [comment]
@gsingh-r7 For the two places we have schema type `array` can we make sure the items type is defined as well:
```
          schema:
            type: array
            items:
              type: string
```

File: komand/server.py [hunk 3]
```diff
@@ -119,15 +145,18 @@ def action_run(name):
...
-                - in: path
-                  name: name
+                - name: name
+                  in: path
                   description: Name of the action
                   type: string
+                - name: input
+                  in: formData
+                  description: Action input
+                  schema: ActionTriggerInputSchema
```
Comment (zyoutz-r7) [inline komand/server.py]: I believe for input we may want to reference the `ActionTriggerInput` definition with a body parameter:
```
  '/actions/{name}':
    post:
      summary: Run an action
      description: Run an action
      parameters:
        - name: name
          in: path
          description: Name of the action
          type: string
          required: true
        - in: body
          name: Action Input
          description: Input to run an action
          schema:
            $ref: '#/definitions/ActionTriggerInput'
```

File: komand/server.py (hunk 2 above)
Comment (gusingh-r7) [reply komand/server.py]: `response` here is of type `<class 'flask.wrappers.Response'>` and contains below:

`{'headers': Headers([('Content-Type', 'application/json'), ('Content-Length', '267')]), '_status_code': 200, '_status': '200 OK', 'direct_passthrough': False, '_on_close': [], 'response': [b'{"body":{"log":"rapid7/Python 3 Script:2.0.1. Step name: run\\nInput: (below)\\n\\n{\'name\': \'gur\'}\\n\\nFunction: (below)\\n\\ndef run(params={}):\\n return {\'name\': params[\'name\'] }\\n\\n","meta":{},"output":{"name":"gur"},"status":"ok"},"type":"action_event","version":"v1"}\n']}`

File: komand/server.py [hunk 4]
```diff
@@ -136,8 +166,14 @@ def action_run(name):
...
-            output = self.run_action_trigger(input_message)
-            return output
+            response = self.run_action_trigger(input_message)
+            action_response = json.loads(response.response[0])
+            action_output = {
+                "body": action_response.get('body'),
+                "type": action_response.get('type'),
+                "version": action_response.get('version')
+            }
+            return jsonify(ActionTriggerOutputSchema().dump(action_output))
```
Comment (zyoutz-r7) [inline komand/server.py]: I think for the time being, lets keep the logic and response the same for these endpoints.  It still allows us to document them with the added schema definitions as you've included.

**Pull request merged**